src/Raven.Server/Documents/Indexes/MapReduce/MapReduceIndexingContext.cs

```diff
@@ -34,24 +37,40 @@ static MapReduceIndexingContext()
 
         public void Dispose()
         {
-            StoreNextMapResultId();
-            DocumentMapEntries?.Dispose();
-            DocumentMapEntries = null;
-            MapPhaseTree = null;
-            ReducePhaseTree = null;
-            ProcessedDocEtags.Clear();
-            ProcessedTombstoneEtags.Clear();
-            StoreByReduceKeyHash.Clear();
-            FreedPages.Clear();
+            try
+            {
+                StoreNextMapResultId();
```

Comment:
I think we talked about moving it to the end of the reduce phase. Why not to call it here:

https://github.com/ravendb/ravendb/blob/5951c3593455d7a42024f538bf30183bfc0c2900/src/Raven.Server/Documents/Indexes/MapReduce/ReduceMapResultsBase.cs#L149-L151

?

Reply:
The problem with calling `StoreNextMapResultId();` here, in `Dispose()`, is that throwing here might hide the previous exception if it was already thrown. In general it's bad idea to have `Dispose` method that can throw I think.

If we moved it next to `WriteEtags` it would be called only if we complete the batch, without any unexpected errors, and that's when we actually want to store `NextMapResultId` (otherwise transaction won't be committed anyway)

Reply:
Done